### PR TITLE
[RW-8670][risk=no] Full Text Search - Add endsWith to findCriteriaAutoComplete call 

### PR DESCRIPTION
**Pull request by chenchals**
Description:

- Added endsWith term capability to full_text search for *findCriteriaAutoComplete* call

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have added explanatory comments where the logic is not obvious
- [x] I have run and tested this change locally, and my testing process is described here
- [x] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java [hunk 1]
```diff
@@ -59,21 +61,50 @@ public MapSqlParameterSource getParameters() {
...
   private static final String DYNAMIC_SQL = "upper(name) like upper(%s)";
 
   private static final String LIMIT_OFFSET = "limit %s offset %s\n";
 
   private static final String OR = "\nor\n";
 
+  private static final String BIND_VAR_DOMAIN = "domain";
+  private static final String BIND_VAR_STANDARD = "standard";
+  private static final String BIND_VAR_TERM = "term";
+  private static final String BIND_VAR_TYPE = "type";
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:93]: I would move these to the very top of the class, so they can be used in the static sql statements as well.

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java [hunk 2]
```diff
@@ -59,21 +61,50 @@ public MapSqlParameterSource getParameters() {
           + "and (%s)\n"
           + "order by est_count desc, name asc\n";
 
+  private static final String AUTO_COMPLETE_ENDS_WITH_WITHOUT_TERM =
+      "select *\n"
+          + "from %s.cb_criteria\n"
+          + "where type = :type\n"
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:67]: I would use the constants you created in the sql statements as well.
```suggestion
          + "where type = :" + BIND_VAR_TYPE + "\n"
```

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java [hunk 3]
```diff
@@ -59,21 +61,50 @@ public MapSqlParameterSource getParameters() {
...
+          + "and is_standard = :standard\n"
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:68]:
```suggestion
          + "and is_standard = :" + BIND_VAR_STANDARD + "\n"
```

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java [hunk 4]
```diff
@@ -59,21 +61,50 @@ public MapSqlParameterSource getParameters() {
...
+          + "and has_hierarchy = 1\n"
+          + "and match(full_text) against(concat('+[', :domain, '_rank1]') in boolean mode)\n"
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:70]:
```suggestion
          + "and match(full_text) against(concat('+[', :" + BIND_VAR_DOMAIN + ", '_rank1]') in boolean mode)\n"
```

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java [hunk 5]
```diff
@@ -59,21 +61,50 @@ public MapSqlParameterSource getParameters() {
...
+          + "and (%s)\n"
+          + "order by est_count desc, name asc\n";
+
+  private static final String AUTO_COMPLETE_ENDS_WITH_WITH_TERM =
+      "select *\n"
+          + "from %s.cb_criteria\n"
+          + "where type = :type\n"
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:77]:
```suggestion
          + "where type = :" + BIND_VAR_TYPE + "\n"
```

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java [hunk 6]
```diff
@@ -59,21 +61,50 @@ public MapSqlParameterSource getParameters() {
...
+          + "and is_standard = :standard\n"
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:78]:
```suggestion
          + "and is_standard = :" + BIND_VAR_STANDARD + "\n"
```

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java [hunk 7]
```diff
@@ -59,21 +61,50 @@ public MapSqlParameterSource getParameters() {
...
+          + "and has_hierarchy = 1\n"
+          + "and match(full_text) against(concat(:term, '+[', :domain, '_rank1]') in boolean mode)\n"
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:80]:
```suggestion
          + "and match(full_text) against(concat(:" + BIND_VAR_TERM + ", '+[', :" + BIND_VAR_DOMAIN + ", '_rank1]') in boolean mode)\n"
```

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java [hunk 8]
```diff
@@ -108,6 +175,14 @@ protected QueryAndParameters generateQueryAndParameters(
                 new BadRequestException(
                     String.format("CDR version with ID %s not found", cdrVersionId)));
 
+    MapSqlParameterSource parameters = new MapSqlParameterSource();
+    Arrays.stream(params)
+        .forEach(
+            param -> {
+              parameters.addValue(param[0].toString(), param[1]);
+            });
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:183]:
```suggestion
        .forEach(
            param -> parameters.addValue(param[0].toString(), param[1]));
```

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/CohortBuilderServiceImpl.java [hunk 9]
```diff
@@ -252,6 +252,46 @@ public List<Criteria> findCriteriaAutoComplete(
...
+
+    Map<String, String> parsedTerms = modifyTermMatchUseEndsWithTerms(term);
+    String modifiedSearchTerm = parsedTerms.get(MODIFIED_TERMS);
+    List<String> endsWithTerms = new ArrayList<>();
+    if (parsedTerms.get(ENDS_WITH_TERMS).length() > 0) {
+      endsWithTerms.addAll(Arrays.asList(parsedTerms.get(ENDS_WITH_TERMS).split(",")));
+    }
+
+    List<DbCriteria> criteriaList = new ArrayList<>();
+    if (endsWithTerms.isEmpty() && modifiedSearchTerm.length() > 0) {
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/CohortBuilderServiceImpl.java:269]:
```suggestion
    if (endsWithTerms.isEmpty() && !modifiedSearchTerm.isEmpty()) {
```

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/CohortBuilderServiceImpl.java [hunk 10]
```diff
@@ -252,6 +252,46 @@ public List<Criteria> findCriteriaAutoComplete(
...
+      criteriaList =
+          cbCriteriaDao.findCriteriaByDomainAndTypeAndStandardAndFullText(
+              domain, type, standard, modifiedSearchTerm, pageRequest);
+    } else if (!endsWithTerms.isEmpty() && modifiedSearchTerm.isEmpty()) {
+      criteriaList =
+          cbCriteriaDao.findCriteriaByDomainAndTypeAndStandardAndNameEndsWith(
+              domain, type, standard, endsWithTerms, pageRequest);
+    } else if (!endsWithTerms.isEmpty() && modifiedSearchTerm.length() > 0) {
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cohortbuilder/CohortBuilderServiceImpl.java:277]:
```suggestion
    } else if (!endsWithTerms.isEmpty() && !modifiedSearchTerm.isEmpty()) {
```

**Review by freemabd — COMMENTED** [review]
Overall it looks good. Requesting a few small changes.

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java (hunk 1 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:93]: done, also pulled other SQL snip statics up.

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java (hunk 3 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:68]: done

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java (hunk 4 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:70]: done

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java (hunk 2 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:67]: done

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java (hunk 5 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:77]: done

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java (hunk 6 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:78]: done

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java (hunk 7 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:80]: done

File: api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java (hunk 8 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cdr/dao/CustomCBCriteriaDaoImpl.java:183]: done

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/CohortBuilderServiceImpl.java (hunk 9 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/CohortBuilderServiceImpl.java:269]: done

File: api/src/main/java/org/pmiops/workbench/cohortbuilder/CohortBuilderServiceImpl.java (hunk 10 above)
Comment (chenchals) [reply api/src/main/java/org/pmiops/workbench/cohortbuilder/CohortBuilderServiceImpl.java:277]: done

**Review by freemabd — APPROVED** [review]
Deployed locally and tested. Looks good.

**Pull request merged**